app/controllers/api/admin/chemicals_controller.rb

```diff
@@ -0,0 +1,238 @@
+module Api
+  module Admin
+    class ChemicalsController < ApiController
+      before_action :authenticate_user!
+      before_action :set_up_errors
+      before_action :fetch_chemical, only: %i[show show_map update names delete_map create_map update_map]
+
+      # GET    /admin/chemicals
+      def index
+        respond_with Chemical.order_by(:id).all
+      end
+
+      # GET   /admin/chemicals/:chemical_id
+      def show
+        render_not_found and return unless @chemical.present?
+
+        respond_with @chemical
+      end
+
+      # POST   /admin/chemicals
+      def create
+        status = :created
+
+        # params = {"data"=>{"attributes"=>{"cas"=>"9-99-9", "name"=>"Kyber (crystaline form)"}, "type"=>"chemicals"}}
+        chemical_params = params['data']['attributes']
+
+        chemical = Chemical.new do |c|
+          c.cas = chemical_params['cas']
+        end
+
+        Sequel::Model.db.transaction do
+          chemical.save raise_on_failure: false
+          begin
+            chemical_name = chemical.reload.add_chemical_name name: chemical_params['name']
+          rescue Sequel::ValidationFailed
+            @errors << chemical_name.stringify_sequel_errors
+            status = :unprocessable_entity
+          end
+        end
+
+        respond_with chemical, status: status
+      end
+
+      # PATCH   /admin/chemicals
+      # PUT     /admin/chemicals
+      def update
+        status = :accepted
+        unless params[:cas].present? || params[:chemical_name_id].present?
+          @errors << "Missing paramter, one is required: cas: #{@chemical&.cas}, chemical_name_id: #{params[:chemical_name_id]}"
+          status = :bad_request
+        end
+
+        Sequel::Model.db.transaction do
+          if params[:cas].present?
```

Comment:
@mcumcu no need to check if param is present, it's already done in L48

Reply:
L48 checks if either param is present. Here we need to check specifically for `params[:cas]`, and update if it exists